tree/dataframe/inc/ROOT/RDF/RMergeableValue.hxx

```diff
@@ -0,0 +1,128 @@
+#ifndef ROOT_RDF_RMERGEABLEVALUE
+#define ROOT_RDF_RMERGEABLEVALUE
+
+#include <functional>
+#include <memory>
+
+#include "RtypesCore.h"
+#include "TList.h"
+
+namespace ROOT {
+namespace RDF {
+
+template <typename T>
+class RMergeableValue {
+protected:
+   T fValue;
+
+public:
+   RMergeableValue(const T &value) : fValue{value} {}
+   virtual ~RMergeableValue() = default;
+   virtual void Merge(const std::unique_ptr<RMergeableValue<T>> &) = 0;
+   virtual const T &GetValue() const { return fValue; }
+};
+
+// Specialization for generic action
+template <typename T>
+class RMergeableGenericAction final : public RMergeableValue<T> {
+   using mergefun_t = std::function<T(const T &, const T &)>;
+   const mergefun_t fMergeFun;
+
+public:
+   RMergeableGenericAction(const T &value, const mergefun_t &mergefun) : RMergeableValue<T>(value), fMergeFun{mergefun}
+   {
+   }
+
+   void Merge(const std::unique_ptr<RMergeableValue<T>> &other) final
+   {
+      const auto &othervalue = other->GetValue();
+      this->fValue = fMergeFun(this->fValue, othervalue);
+   }
+};
+
+// RMergeableValue specialization for the Mean action
+template <typename T>
+class RMergeableMean final : public RMergeableValue<T> {
+   ULong64_t fCounts;
+
+public:
+   RMergeableMean(const T &value, const ULong64_t &counts) : RMergeableValue<T>(value), fCounts{counts} {}
+
+   const ULong64_t &GetCounts() const { return fCounts; }
```

Comment:
it's standard practice to just return fundamental types by value (it's slightly more performant)